modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/watch/WatchAggregator.java

```diff
@@ -0,0 +1,243 @@
+     */
+    public Optional<AggregatedWatch> watch(long revision, BiConsumer<Collection<IgniteBiTuple<Key, byte[]>>, Long> saveRevisionAct) {
+        synchronized (watches) {
+            if (watches.isEmpty())
+                return Optional.empty();
+            else
+                return Optional.of(new AggregatedWatch(inferGeneralCriteria(), revision, watchListener(saveRevisionAct)));
+        }
+    }
+
+    /**
+     * Returns general criterion, which overlays all aggregated criteria.
+     *
+     * @return aggregated criterion.
+     */
+    // TODO: We can do it better than infer range always
+    private KeyCriterion inferGeneralCriteria() {
+        return new KeyCriterion.RangeCriterion(
+            watches.values().stream().map(w -> w.keyCriterion().toRange().getKey()).min(Key::compareTo).get(),
+            watches.values().stream().map(w -> w.keyCriterion().toRange().getValue()).max(Key::compareTo).get()
+        );
+    }
+
+    /**
+     * Produces the watch listener, which will dispatch events to appropriate watches.
+     *
+     * @param storeRevision action to commit keys-revision pair to persistent store for processed keys.
+     * @return watch listener, which will dispatch events to appropriate watches.
+     */
+    private WatchListener watchListener(BiConsumer<Collection<IgniteBiTuple<Key, byte[]>>, Long> storeRevision) {
+        // Copy watches to separate collection, because all changes on the WatchAggregator watches
+        // shouldn't be propagated to listener watches immediately.
+        // WatchAggregator will be redeployed with new watches if needed instead.
+        final LinkedHashMap<Long, Watch> cpWatches = new LinkedHashMap<>(watches);
+
+        return new WatchListener() {
+
+            @Override public boolean onUpdate(@NotNull Iterable<WatchEvent> evts) {
+                var watchIt = cpWatches.entrySet().iterator();
+                Collection<Long> toCancel = new ArrayList<>();
+
+                while (watchIt.hasNext()) {
+                    Map.Entry<Long, WatchAggregator.Watch> entry = watchIt.next();
+                    WatchAggregator.Watch watch = entry.getValue();
+                    var filteredEvts = new ArrayList<WatchEvent>();
+
+                    for (WatchEvent evt : evts) {
+                        if (watch.keyCriterion().contains(evt.oldEntry().key()))
+                            filteredEvts.add(evt);
+                    }
+
+                    if (!filteredEvts.isEmpty()) {
+                        if (!watch.lsnr().onUpdate(filteredEvts)) {
+                            watchIt.remove();
+
+                            toCancel.add(entry.getKey());
+                        }
+                    }
+                }
+
+                // Cancel finished watches from the global watch map
+                // to prevent finished watches from redeploy.
+                if (!toCancel.isEmpty())
+                    cancelAll(toCancel);
```

Comment:
How the aggregated watch is redelpoyed on some sub watches cancellation ?

Reply:
Redeploy initiated from inside the listener itself will be implemented under https://issues.apache.org/jira/browse/IGNITE-14604 soon. At the moment, in the worst case, we will listen to redundant keys, but will not pass them to any watches.